### PR TITLE
Finish the project context implementation for the free-threaded project system shims

### LINKED ISSUE #31019: IWorkspaceProjectContext.Dispose() still has UI thread affinity
@davkean still discovered it's got one icky UI thread requirement here:

```
    Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.Contract.ThrowIfFalse(bool condition, string message)    Unknown
    Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.ForegroundThreadAffinitizedObject.AssertIsForeground()    Unknown
    Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.SetDocumentContext(Microsoft.CodeAnalysis.DocumentId documentId)    Unknown
    Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.ClearOpenDocument(Microsoft.CodeAnalysis.DocumentId documentId, bool isSolutionClosing)    Unknown
    Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.ClearOpenDocuments(Microsoft.CodeAnalysis.ProjectId projectId)    Unknown
    Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.ClearProjectData(Microsoft.CodeAnalysis.ProjectId projectId)    Unknown
    Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.OnProjectRemoved(Microsoft.CodeAnalysis.ProjectId projectId)    Unknown
    Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.OnProjectRemoved(Microsoft.CodeAnalysis.ProjectId projectId)    Unknown
    Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.RemoveFromWorkspace.AnonymousMethod__86_0(Microsoft.CodeAnalysis.Workspace w)    Unknown
    Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyChangeToWorkspace(System.Action<Microsoft.CodeAnalysis.Workspace> action)    Unknown
    Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.RemoveFromWorkspace()    Unknown
    Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.CPS.CPSProject.Dispose()    Unknown
```

### PR DESCRIPTION
**Pull request by jasonmalinowski**
The handling for determination of project contexts wasn't finished when we merged the work for free-threaded project system shims. This completes the work and fixes a bunch of issues as well.

This should fix the Roslyn side of #21358, but there are also some project system bugs that have been found too that will impact it.

Reviewers are **strongly** encouraged to go commit-by-commit here.

Work remaining:

- [x] Remove the existing hack around skipping locks when we remove projects, since those don't work correctly if just files are removed.

<details><summary>Ask Mode template</summary>

### Customer scenario

The customer has a multitargeting SDK project, linked files, or is using shared asset projects. If they try to use the dropdown in the editor to change the context, it won't work reliably or at all.

### Bugs this fixes

#31019
#31365
#21358 (but still needs more fixes from the Visual Studio side, it seems)

### Workarounds, if any

Only workaround would be to use 15.9.

### Risk

Moderate: this is a technical area in the code and interacts with a number of Visual Studio components. In the process of writing this we've discovered at least three bugs in _other_ components, to give a sense of how tricky this area is!

### Performance impact

Better; this also fixes a few memory leaks that would have leaked some project information when projects and solutions were closed.

### Is this a regression from a previous update?

Yes.

### Root cause analysis

This was a known limitation when we merged the original free-threaded project initialization code. This only impacts a subset of our users, and when they do hit it is supremely annoying, but it was best to get all the other stuff in their hands and worry about this later. The existing code already had some bugs too; since we were rewriting all of it meant it was a good opportunity to get it right.

### How was the bug found?

Various bugs reported from customers, but this was tracked internally in our own backlog.

</details>

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
Also @DustinCampbell you might find this PR interesting; it's cleaning up a bit of the document context stuff, and I think it looks reasonable enough that the internal bits could be made public. There's also one case where I obsoleted a method that took a parameter that never should have been there in the first place, but if we think nobody is calling it I'm OK with an API break to just delete it.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs [hunk 1]
```diff
@@ -207,66 +207,69 @@ private ProjectId GetActiveContextProjectId(IVsHierarchy hierarchy, IEnumerable<
...
+                }
 
-                    if (matchingProjectId != null)
-                    {
-                        return matchingProjectId;
-                    }
+                // We may have multiple projects with the same hierarchy, but we can use __VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext to distinguish
+                if (ErrorHandler.Succeeded(hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext, out object contextProjectNameObject)))
+                {
+                    EnsureSubscribedToHierarchyPropertyChanges(hierarchy);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs]: ❔ Why is this only inside the `ErrorHandler.Succeeded` guard? The intermediate hierarchies seem to get monitored unconditionally.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs [hunk 2]
```diff
@@ -207,66 +207,69 @@ private ProjectId GetActiveContextProjectId(IVsHierarchy hierarchy, IEnumerable<
...
-                            }
-                        }
-                    }
+                if (matchingProjectId != null)
+                {
+                    return matchingProjectId;
                 }
 
                 // If we had some trouble finding the project, we'll just pick one arbitrarily
                 return projectIds.First();
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs]: ❔ Do we not want to use Miscellaneous Files for this?

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs]: Not really. projectIds in this case is a list of projects that already contains this file, presumably they have the right references, or at least much better references than Miscellaneous Files would provide.

This is really the "last resort" call; I should add a comment clarifying that we really don't ever want to end up here, but if anything goes wrong with another project system, we can't entirely blow up.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs]: (Warning: GitHub is flagging your comment as outdated; it's possible you're commenting on an older commit that I might have already fixed.)

The intermediate ones are only getting watched if they actually gave a pointer to another IVsHierarchy. Basically, you can think of this code as taking an IVsHierarchy and following pointers to other IVsHierarchies or projects. We're only sticking watchers on IVsHierarchies that gave us a redirection, so we'll know if they redirect elsewhere.

**Pull request merged**